src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,395 @@
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+};
+
+//! Calculate cost over batches.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
+                                          const size_t begin,
+                                          const size_t batchSize)
+{
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_cols)
+    Projection(iterate);
+
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
+
+  for (size_t i = begin; i < begin + batchSize; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+         double eval = metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(targetNeighbors(j, i))) -
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+}
+
+//! Compute gradient over whole dataset.
+template<typename MetricType>
+template<typename GradType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& /* coordinates */,
+                                        GradType& gradient)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  arma::mat cij = arma::zeros(dataset.n_rows, dataset.n_rows);
+  arma::mat cil = arma::zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      cij += diff * arma::trans(diff);
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate gradient due to triplets.
+        arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        cil += diff * arma::trans(diff);
+
+        diff = dataset.col(i) - dataset.col(impostors(l, i));
+        cil -= diff * arma::trans(diff);
+      }
+    }
```

Comment:
When I reviewed this the first time I thought it was okay, but I realized I forgot about the `std::max(0.0, 1 + eval)`, and I don't think that `max()` is accounted for here.  I *think* that we can say that if the constraint is satisfied, the gradient will be 0 and we can ignore it.  I guess the condition would be `eval < -1`.  I am not 100% sure on the derivation there, but I think it is at least the right direction.

Reply:
Hmm, This looks like a valid point. But I don't think the authors have considered this. I think we can try imposing this condition. 
>  I guess the condition would be eval < -1

This condition seems reasonable. So, it means we will be calculating eval once again, here in gradient. Right?

Reply:
Ack, sorry, I read the `[ ... ]_+` from the evaluation bit.  Sorry for the error on my part.

But note that the gradient computation (the equation below (18)... why isn't it numbered?) does not use every possible tuple.  Instead it only uses some:

"Let us define a set of triples N_t, such that (i, j, l) ∈ N_t if and only if the indices (i, j, l) trigger the hinge loss in the second part of Eq. (18)."

So I do think we have to compute `eval` again in order to find the set of triples N_t that we need to use for the gradient.

Reply:
> Let us define a set of triples N_t, such that (i, j, l) ∈ N_t if and only if the indices (i, j, l) trigger the hinge loss in the second part of Eq. (18).

Right, this line contains the thing. Okay, I will make the required change and will look into the results.

Reply:
@rcurtin, this was the exact problem. Now, I ran quite a number of tests on multiple dataset and everything looks promising. :) 
Here is a result: [link](https://pasteboard.co/HnB3o5x.png)